### PR TITLE
Remove trailing white space and add CI check

### LINKED ISSUE #1864: Need linting CI check to prevent trailing whitespace
There are a lot of lines with trailing whitespace. @pyvista/developers please make sure to configure your IDE's settings to trim trailing whitespace when saving files - this is a common practice

Fore example:

https://github.com/pyvista/pyvista/blob/829983117b3b1cf44e5ac00bb0dc16d6436ed391/pyvista/core/filters/data_set.py#L327

We need to implement some sort of linting CI check to prevent bad code formatting like this. `flake8` is a great way to go

### PR DESCRIPTION
**Pull request by banesullivan**
### Overview

This PR trims out all trailing whitespace and adds [W291](https://www.flake8rules.com/rules/W291.html) to the stylecheck routine.

I want this to land as soon as possible but I think in a follow up PR, we should expand that flake8 check to catch other things

Resolve #1864

### CODE REVIEW

File: requirements_style.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
-codespell==2.1.0
-pydocstyle==6.1.1
+codespell>=2.1.0
+pydocstyle>=6.1.1
```
Comment (banesullivan) [inline requirements_style.txt:2]: Not sure why these were pinned... freeing them seems to have had no consequences and we should always try to use the latest versions for these sorts of things (because codespell's dictionary is regularly updated)

**Comment by banesullivan** [comment]
The CI robots are happy so I am going to use my admin powers to merge this

File: requirements_style.txt (hunk 1 above)
Comment (tkoyama010) [reply requirements_style.txt:2]: I added the issue https://github.com/pyvista/pyvista/issues/1868 . Thank you for pointing our.

**Comment by adeak** [comment]
May I ask what the urgency was that got this PR with 136 diffs across 35 files merged without review?

**Comment by banesullivan** [comment]
Trailing white space is a common and big enough of a problem that almost every editor/IDE has functionality to remove it automatically, and most styling best standards require you to eliminate all instances of it (reference [NumPy's style guidelines](https://github.com/numpy/numpy/blob/b892ed2c7fa27b2e0d73c12d12ace4b4d4e12897/doc/neps/nep-0045-c_style_guide.rst#code-layout))

While this was not outlined in our `CONTRIBUTING.md`, it should be... I suppose I assumed this to be a universal standard when I wrote the contributing guidelines.

Furthermore, this has come up as an issue time and time again:

- https://github.com/pyvista/pyvista/pull/474#discussion_r350859610
- https://github.com/pyvista/pyvista/pull/805#discussion_r443089735
- https://github.com/pyvista/pyvista/pull/1334#discussion_r637456494 / https://github.com/pyvista/pyvista/pull/1334#issuecomment-849840259
- Something weird you pointed out: https://github.com/pyvista/pyvista/pull/1428#issue-928046195
- White space changes needlessly showing up in people's PRs:
  - https://github.com/pyvista/pyvista/pull/1593/commits/1ce5c8d3187bd9f3b03579a5a0263f3fbc5f51dd
  - https://github.com/pyvista/pyvista/pull/1851/files#diff-c657f41811e99e84bf27a549ecf493b113546c6c0f925cdb5f7cc85c8683e615
  - etc.!

The problem I see with trailing white space, while there are many -- google ["why is trailing white space bad?"](https://www.google.com/search?q=why+is+trailing+whitespace+bad%3F&oq=why+is+trailing+whitespace+bad%3F), is the majority of developer's editors will automatically trim it. This leads to diffs in actual code-changing PRs that are bigger than they need to be, and if two developers are working in the same file, their changes will conflict because they both trimmed the same white space -- this is an absolute nuisance. The opposite case also holds for developers who do not trim it - their diffs are still needlessly bigger. In both cases, making PRs more cumbersome to review.

> May I ask what the urgency was

I am working locally on several branches to resolve outstanding issues and contribute while I have free time. My time is limited and dealing with selecting chunks to stage and commit that avoid these white space areas was pressing enough that I wanted it in `main` so I could rebase my local branches to avoid needless conflicts and wasted time handling this.

> PR with 136 diffs across 35 files merged

"X diffs and Y files" makes this sound like much bigger of a change than it actually is. FWIW, I did not adjust a single character of actual code - I removed *only trailing white space*, and put a CI check in place to prevent this moving forward. So this should only have very minimal conflict with any existing PRs where this white space was adjusted. It would be more challenging to deal with conflicts if two actual code-changing PRs altered the white space. In any case, perhaps I'm the only developer here who uses an editor that automatically trims the white space -- then this should have no effect unless someone happened to edit the exact lines I trimmed the white space from in a currently open PR, which would still be minimal.

> ... without review?

I understand your concerns about merging a pull request without review, and I want to reassure you that this is not something I take lightly with PyVista. The pressing nature of this for the sake of my time to make additional contributions posed enough of a benefit to outweigh the consequences of merging without review. Because I did not change any actual code and because I have very strong opinions on this matter, I made the executive decision to merge this.

While my contributions to PyVista have been limited over the last year due to new obligations for my day job and personal matters that have occupied my free time/led to a bit of burn out, I do plan to ramp up my contributions to PyVista moving forward form here. This pull request was an action taken to foster that. @adeak, @akaszynski,@tkoyama010, 
 and @GuillaumeFavelier: if my actions here were out of line, please let me know so that I can be even more cautious moving forward.

**Comment by GuillaumeFavelier** [comment]
@banesullivan for trailing white spaces it's okay for me :+1:

File: requirements_style.txt (hunk 1 above)
Comment (akaszynski) [reply requirements_style.txt:2]: > Not sure why these were pinned... freeing them seems to have had no consequences and we should always try to use the latest versions for these sorts of things (because codespell's dictionary is regularly updated)

The reason why I've pinned these is when a new package is released, our PRs have the potential to suddenly fail. This means that new or existing contributors may find their PRs failing for some reason they don't quite understand and pinning the packages helps alleviate this issue.

**Comment by adeak** [comment]
Thanks for the reply @banesullivan. I'll be the first to tell you that I'm unreasonably pedantic about sticking to best practices and good style. Which is why I never questioned why these mistakes are great to fix.

> I am working locally on several branches to resolve outstanding issues and contribute while I have free time. My time is limited and dealing with selecting chunks to stage and commit that avoid these white space areas was pressing enough that I wanted it in `main` so I could rebase my local branches to avoid needless conflicts and wasted time handling this.

I understand. I would expect that rebasing would handle 99% of cases, but perhaps these lines are more prevalent than I'd think.

> So this should only have very minimal conflict with any existing PRs where this white space was adjusted.

Yes, I'm not concerned about clashes with other PRs. In fact I figured that there would be very little conflict with other PRs, which is why I didn't understand the rush; there would be no harm in getting other PRs open while this one was undergoing review. (But you've explained the rush.)

> I understand your concerns about merging a pull request without review, and I want to reassure you that this is not something I take lightly with PyVista.

This is the only aspect I have issue with (which is why I phrased my question the way I did). It's not without precedent to have things merged without review in an executive decision, but it normally happens when there's an imminent release we want to have a PR in, and everyone's busy so the PR's been open a few days without an approval.

>  if my actions here were out of line, please let me know so that I can be even more cautious moving forward.

Well, I would say that it's best not to merge PRs without review (29 minutes after the PR was opened), but you already agree with that. The PR's changes as you hinted above are probably no-brainers: whitespace here and there and a version pinning change. But people have a tendency to assume they do everything right, and code review is exactly the mechanic that prevents hubris from affecting the quality of a codebase :) Especially when there's a frenzy of contributions it's easy to make mistakes.

In this case, the changes look good to me. _However_, had I had a chance to review the code I would have suggested to separate the whitespace removal and the pinning change into two separate PRs. If we could get the styling changes merged in a single commit, we could have [used an ignore revs file to make git blame ignore this commit going forward](https://git-scm.com/docs/git-config#Documentation/git-config.txt-blameignoreRevsFile). Not having style changes show up in `git blame` would be nice. Since the pinning change is slapped on to this PR's merge commit, this is a no-go and we're stuck with these changes showing up in the blame.

None of this is a huge deal, just a tempest in a teapot. But it might show that even with no-brainer changes it's possible that any one user's insight is suboptimal compared to what we can put together if we pool our resources. All that being said, there's no real harm done, and what matters is that you can use your free time well to contibute to PyVista. I'm certainly in no position to hold anyone accountable for their contibutions :) Cheers.

**Comment by akaszynski** [comment]
**tl;dr**

- [ ] Let's make a PR that introduces a flake8 style check.
- We're all doing a great job here; we just need to remember that there are a variety of approaches for writing good software and we're all in it together.

---
> I'm unreasonably pedantic about sticking to best practices and good style

I view this as a positive thing. In any open source project, there should be adherence to some minimum styling standard. One way we can enforce this is by using a style standard like `flake8`. If the CI robots approve, then we don't need to argue about it; we have a gatekeeper and everyone likes green checkmarks.

> Well, I would say that it's best not to merge PRs without review (29 minutes after the PR was opened), but you already agree with that. The PR's changes as you hinted above are probably no-brainers: whitespace here and there and a version pinning change. But people have a tendency to assume they do everything right, and code review is exactly the mechanic that prevents hubris from affecting the quality of a codebase :) Especially when there's a frenzy of contributions it's easy to make mistakes.

@adeak has a point. Both @banesullivan and I have pushed a ton of code to this repository, and this repo even used to be named (a long long time ago) `akaszynski/vtki`. It was a difficult move, but I decided that to get more contributions and visibility, @banesullivan convinced me that it would be best to create a new open source organization that would accept contributions from all and we would assume the role of "admins" for this organization. My hope is for both @banesullivan and I to manage the tone, style, and release cadence without dominating the repository as we genuinely welcome contributions from all (and it's fun to see it doing so well).

Given that all the contributors here have day jobs and no one is getting a cent from our work, the end goal for this library has always been:
- Be a useful library that extends VTK
- Be open to new and existing contributors
- Foster a community where people feel free to post issues, PRs, and ask discussions without recourse.

How that ends up being executed means balancing `import this`, which is often conflicting and contradictory. However, that's software development. One thing we should remember is no one is under the gun to get a release out, and while I often impose release deadlines, that's often just to keep us "moving" so the users of `pyvista` can get the latest and greatest features (albeit sometimes bugs).

We all bring something unique to the table, and I must constantly remind myself that this isn't just my repository, but now the communities; I gave up ownership years ago when @banesullivan and I co-created `pyvista`. It's not easy, and if I had a dollar for everytime I wanted to just merge when the checks were green I wouldn't have to work for my day job. However, I feel that I've grown a lot from having to take a step back and obtain agreement (sometimes begrudging) with those who have volunteered their time and effort to review my (sometimes half-baked) PRs.

At the end of the day, provided we can balance creating a useful library while balancing our unique skills and personalities, we've done our job. I don't think anyone who's contributed has ever though otherwise, and I'm honestly impressed with the github community that such a thing is possible.

**Comment by banesullivan** [comment]
> This is the only aspect I have issue with (which is why I phrased my question the way I did)

Apologies, I understood that. When I realized that others might have an actual concern with me merging this particular change without review I felt I needed to add justification for the change as my original description at the top doesn't make a strong case.

Moving forward, I will continue to use my ability to merge without review very sparingly. Thanks for sharing the concern about freeing the pinned dependencies and changing style in the same commit -- I did not consider this and did not know that a commit for stylistic changes could be ignored; that's quite neat.

> But it might show that even with no-brainer changes it's possible that any one user's insight is suboptimal compared to what we can put together if we pool our resources.

Very fair! I'm hoping this particular change is minimal enough that I didn't miss too much of an opportunity to handle this in a tangibly better fashion. However, sometimes I want to move fast (and cut corners). The engineering methods can be an burdensome delay when trying to get things done -- but like I've hopefully conveyed by now, this isn't something I take lightly and do not plan to merge PRs without review but in very specific cases that I either feel incredibly strongly about or can take full responsibility for.

File: requirements_style.txt (hunk 1 above)
Comment (banesullivan) [reply requirements_style.txt:2]: Ah I see. I'd like to keep them unpinned though if possible - the fact that they are in their own GitHub action seperate from the actual tests should make this easy enough to deal with (in theory)

If these fail again ping me and I'll jump on it 😉

**Comment by tkoyama010** [comment]
Personally, I really like PyVista's policy of blocking unreviewed PR. However, maintainer has limited resources and may not always be able to review. Therefore, I think there is no problem with merging past PRs that have obvious defects or do not affect the program content, as this is a privilege given to the maintainers. In that case, post-merge confirmation is very important, as @adeak asked the question and @banesullivan answered it. If we discover a problem at that point, let's create a revert pull request. (Challenges within privileges is always welcome, and revert due to failure is nothing to be ashamed of.)
